### PR TITLE
feat: Produce attribution data to the kafka topic

### PR DESCRIPTION
**Pull request by evanh**
This will produce attribution data to a new Kafka topic, ultimately to be
consumed by a super-big-consumer and published to PubSub. This collects all the
bits of data we care about, puts them into a new container and publishes it to
Datadog and Kafka.

This topic has already been created by the ops team.

### CODE REVIEW

File: snuba/attribution/log.py [hunk 1]
```diff
@@ -33,3 +49,42 @@ def log_attribution(attr_data: AttributionData) -> None:
...
+        kfk.poll(0)  # trigger queued delivery callbacks
+        kfk.produce(
+            settings.KAFKA_TOPIC_MAP.get(
+                Topic.ATTRIBUTION.value, Topic.ATTRIBUTION.value
+            ),
+            data_str.encode("utf-8"),
+            on_delivery=_record_attribution_delivery_callback,
+        )
+    except Exception as ex:
+        logger.exception("Could not record attribution due to error: %r", ex)
```
Comment (nikhars) [inline snuba/attribution/log.py:90]: I would be more comfortable by setting this to logger.warning. If we don't log attribution for some reason, we don't want to get alerted/paged about it.

File: snuba/attribution/log.py [hunk 2]
```diff
@@ -33,3 +49,42 @@ def log_attribution(attr_data: AttributionData) -> None:
             "table": q.table,
         }
         metrics.increment("log", q.bytes_scanned, tags=tags)
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/attribution/log.py:51]: 1) Number of queries can be inferred by the size of the `queries` array.
2) Attribution is not meant for debugging. It sounds like we need to add a dataset flag to the querylog data.

File: snuba/attribution/__init__.py [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
+
 from snuba import environment, state
 from snuba.utils.metrics.wrapper import MetricsWrapper
 
 from .appid import AppID
 from .default import APPIDS as DEFAULT_APPIDS
 
 metrics = MetricsWrapper(environment.metrics, "snuba.attribution")
-
+logger = logging.getLogger("snuba.attribution")
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/attribution/__init__.py:12]: Yeah I'll remove. This was for an earlier iteration of this that didn't work.

[the remark this answers is not in the document]
Comment (evanh) [reply snuba/attribution/__init__.py:12]: This is used in the `log.py` file to record the warnings etc.

**Comment by lynnagara** [comment]
Can we make sure we flush this producer when we shutdown processes? Otherwise we will have the same problem as we do today with the querylog producer. During shutdown we don't flush the producer so processes don't get closed properly and we fail to record all queries during shutdown.

File: snuba/cli/subscriptions_scheduler_executor.py [hunk 4]
```diff
@@ -139,7 +140,7 @@ def handler(signum: int, frame: Any) -> None:
     signal.signal(signal.SIGINT, handler)
     signal.signal(signal.SIGTERM, handler)
 
-    with executor, closing(producer), flush_querylog():
+    with executor, closing(producer), flush_querylog(), flush_attribution_producer():
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/cli/subscriptions_scheduler_executor.py]: Done

File: snuba/attribution/log.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (volokluev) [reply snuba/attribution/log.py:53]: I think adding dataset is a good idea

File: snuba/attribution/log.py [hunk 5]
```diff
@@ -33,3 +51,62 @@ def log_attribution(attr_data: AttributionData) -> None:
...
+    try:
+        yield
+    finally:
+        if kfk is not None:
+            messages_remaining = kfk.flush()
+            logger.debug(f"{messages_remaining} querylog messages pending delivery")
+
+
+def record_attribution(attr_data: AttributionData) -> None:
+    global kfk
```
Comment (volokluev) [inline snuba/attribution/log.py:83]: can we put the creation of the attribution producer in its own function?

```python
def _attribution_producer():
       if kfk is None:
            kfk = Producer(build_default_kafka_producer_configuration())
       return kfk

**Pull request merged**